### PR TITLE
preserve subdomains & boundaries in MeshQuad._splitquads

### PR DESCRIPTION
**Pull request by gdmcbain**
For #385

### CODE REVIEW

**Comment by gdmcbain** [comment]
I noticed that `MeshQuad._splitquads` isn't covered by a test.  I'll see if I can add one.

**Comment by kinnala** [comment]
This naming, `_splitquads`, seems to imply that it is a private method. Originally it was, as the reason for introducing it was to do plotting of solutions on quadrilateral meshes and it was used internally by `MeshQuad.plot3`.

Currently I can think of many other use cases so I wonder if we should add an alias, `MeshQuad.split_quads`, `MeshQuad.split`, or something. Note that we also have the variant `MeshQuad._splitquads_symmetric` which is supposed to split each quad into four triangles.

**Comment by gdmcbain** [comment]
I see, yes, good point about the private-looking name.

On the two new names, the 'quads' in the first seems redundant but then there are lots of other functions and methods already called split.  And maybe it could be thought to be an inverse of the new `Mesh.__add__ `, so maybe the former.  Or what about `MeshQuad.to_meshtri`?

And for the symmetric variant, I hoping that they can be refactored later.  As I generally have to work with unstructured meshes, I haven't looked at it myself.

**Comment by gdmcbain** [comment]
Oh, I see that `MeshQuad._splitquads_symmetric` introduces new points in the centroids; that'll break the implementation above.  And the tests, which are based on checking that the points supporting the subdomains and boundaries don't change.  Can that be deferred?

**Comment by kinnala** [comment]
Yes, `to_meshtri` sounds reasonable. NumPy/SciPy uses `tolist`, `toarray`, etc. but I think `tomeshtri` sounds too terse. Also, we already have `to_dict` and `to_json` etc.

We can later add `m.to_meshtri(symmetric=True)` flag.

File: skfem/mesh/mesh2d/mesh_quad.py [hunk 1]
```diff
@@ -244,21 +245,41 @@ def _uniform_refine(self):
 
         self._fix_boundaries(new_facets)
 
-    def _splitquads(self, x=None):
+    def to_trimesh(self, x=None):
```
Comment (kinnala) [inline skfem/mesh/mesh2d/mesh_quad.py:248]: `s/trimesh/meshtri/`

File: tests/test_mesh.py [hunk 2]
```diff
@@ -181,5 +181,21 @@ def runTest(self):
...
+
+        for s in mesh.subdomains:
+            self.assertEqual(np.setdiff1d(*[m.t.T[m.subdomains[s]]
+                                            for m in [mesh, tri_mesh]]).size,
+                             0)
+
+        for b in mesh.boundaries:
+            np.testing.assert_array_equal(*[m.facets.T[m.boundaries[b]]
+                                            for m in [mesh, tri_mesh]])
+
```
Comment (kinnala) [inline tests/test_mesh.py:198]: Here is an additional test that should be resolved:
```python
>>> m = MeshQuad()
>>> m.define_boundary('left', lambda x: x[0] == 0)
>>> m.to_trimesh()
Triangular mesh with 4 vertices and 2 elements.
>>> m.to_trimesh().boundaries
{'left': array([1])}
>>> m.refine()
>>> m.to_trimesh()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/src/scikit-fem/skfem/mesh/mesh2d/mesh_quad.py", line 269, in to_trimesh
    for f in self.facets.T[self.boundaries[k]]])
  File "/src/scikit-fem/skfem/mesh/mesh2d/mesh_quad.py", line 269, in <listcomp>
    for f in self.facets.T[self.boundaries[k]]])
StopIteration
```

File: skfem/mesh/mesh2d/mesh_quad.py (hunk 1 above)
Comment (gdmcbain) [reply skfem/mesh/mesh2d/mesh_quad.py]: 6e326be

File: tests/test_mesh.py (hunk 2 above)
Comment (gdmcbain) [reply tests/test_mesh.py:198]: Ugh.  Is this something to do with #104? That has `MeshQuad` ticked. 

I knew those functional programming people were onto something when they lauded immutability.

Comment (gdmcbain) [reply tests/test_mesh.py:198]: `m.to_meshtri().boundaries['left'] == [6, 1]`, which isn't sorted which was assumed.  I'll sort.  Is it only `refine` that deranges boundaries?  I'll add the test.

It's also assumed that `.facets` is sorted lexicographically.

Comment (gdmcbain) [reply tests/test_mesh.py:208]: I've sorted in _splitquads, but an alternative would be to sort in `Mesh._fix_boundaries`. 

https://github.com/kinnala/scikit-fem/blob/b5571dac8bad50aee2da1ad6a1146f488c49e400/skfem/mesh/mesh.py#L144-L145

Comment (gdmcbain) [reply tests/test_mesh.py:208]: ...instead of `.flatten()`

Comment (kinnala) [reply tests/test_mesh.py:208]: > I knew those functional programming people were onto something when they lauded immutability.

Yeah, it's an old mistake that `refine` does not return a new mesh. Sometimes I ran into an issue where I run `m.refine()` and then the respective `InteriorBasis` breaks.

I wonder if we should do something about it or not.

Comment (kinnala) [reply tests/test_mesh.py:208]: > I've sorted in _splitquads, but an alternative would be to sort in `Mesh._fix_boundaries`.
> 
> https://github.com/kinnala/scikit-fem/blob/b5571dac8bad50aee2da1ad6a1146f488c49e400/skfem/mesh/mesh.py#L144-L145

I cannot currently recall a reason why there couldn't be a sort there.

I'll add an issue as a reminder.

**Pull request merged**